<?xml version="1.0"?>
<launch>
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find racecar_gazebo)/worlds/racecar_runway.world"/>
    <arg name="paused" value="false"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="true"/>
  </include>

  <param name="robot_description" command="$(find xacro)/xacro $(find racecar_description)/urdf/racecar.urdf.xacro"/>
  <node name="spawn_racecar" pkg="gazebo_ros" type="spawn_model" output="screen"
        args="-urdf -param robot_description -model racecar -x 0 -y 0 -z 0.05"/>

  <rosparam file="$(find racecar_gazebo)/config/racecar_control.yaml" command="load"/>
  <node name="controller_manager" pkg="controller_manager" type="spawner" respawn="true" output="screen" ns="/racecar"
      args="left_rear_wheel_velocity_controller right_rear_wheel_velocity_controller
            left_front_wheel_velocity_controller right_front_wheel_velocity_controller
            left_steering_hinge_position_controller right_steering_hinge_position_controller
            joint_state_controller"/>

  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" respawn="false" output="screen">
    <remap from="joint_states" to="/racecar/joint_states"/>
  </node>

  <node name="drive_controller" pkg="racecar_gazebo" type="drive_controller.py" output="screen">
  </node>

  <node name="racecar_odometry" pkg="racecar_gazebo" type="odometry.py" output="screen">
    <param name="robot_name" value="racecar"/>
    <param name="base_frame" value="base_footprint"/>
    <param name="odom_frame" value="odom"/>
  </node>

  <node name="map_server" pkg="map_server" type="map_server" args="$(find racecar_gazebo)/map/racecar_runway.yaml" />

  <include file="$(find racecar_gazebo)/launch/amcl.launch">
    <arg name="scan_topic" value="/scan" />
    <arg name="initial_pose_x" value="0.0"/>
    <arg name="initial_pose_y" value="0.0"/>
    <arg name="initial_pose_a" value="0.0"/>
  </include>

  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <rosparam file="$(find racecar_gazebo)/config/move_base_config/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find racecar_gazebo)/config/move_base_config/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find racecar_gazebo)/config/move_base_config/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find racecar_gazebo)/config/move_base_config/global_costmap_params.yaml" command="load" />
    
    <rosparam file="$(find racecar_gazebo)/config/move_base_config/base_local_planner_params.yaml" command="load" />
    
    <remap from="cmd_vel" to="/cmd_vel"/>
  </node>

  </launch>